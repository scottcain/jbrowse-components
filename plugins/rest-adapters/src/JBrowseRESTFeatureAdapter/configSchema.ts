import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'JBrowseRESTFeatureAdapter',
  {
    location: {
      type: 'fileLocation',
      description: 'the base URL for the REST endpoints of this API',
      defaultValue: {
        uri: 'https://example.com/path/to/my/rest/endpoint',
        locationType: 'UriLocation',
      },
    },
    assemblyNames: {
      type: 'stringArray',
      description:
        'list of assembly names contained in this REST API. only used if the assembly_names optional resource is not implemented',
      defaultValue: [],
    },
    extra_query: {
      description: 'additional URL query values to pass to the REST API',
      type: 'frozen',
      defaultValue: {},
    },
    optional_resources: ConfigurationSchema(
      'JBrowseRESTFeatureAdapterOptionalResources',
      {
        assembly_names: {
          description:
            'set to true if the REST endpoint provides a list of its configured assemblies',
          type: 'boolean',
          defaultValue: true,
        },
        has_data_for_ref: {
          description:
            'set to true if the REST endpoint provides a list of reference sequence names for its configured assemblies',
          type: 'boolean',
          defaultValue: true,
        },
        get_ref_names: {
          description:
            'set to true if the REST endpoint provides a list of reference sequence names for its configured assemblies',
          type: 'boolean',
          defaultValue: false,
        },
        region_stats: {
          description:
            'set to true if the REST endpoint provides region statistics',
          type: 'boolean',
          defaultValue: false,
        },
        region_feature_densities: {
          description:
            'set to true if the REST endpoint provides region feature densities',
          type: 'boolean',
          defaultValue: false,
        },
      },
    ),
  },
  { explicitlyTyped: true },
)
