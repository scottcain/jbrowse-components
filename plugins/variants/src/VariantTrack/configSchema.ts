import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { createBaseTrackConfig } from '@jbrowse/core/pluggableElementTypes'
import PluginManager from '@jbrowse/core/PluginManager'

/**
 * #config VariantTrack
 * Mostly similar to feature track, but has `ChordDisplayType` registered to it,
 * and custom feature details in `LinearVariantDisplay`
 */
export default function (pluginManager: PluginManager) {
  return ConfigurationSchema(
    'VariantTrack',
    {},
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: createBaseTrackConfig(pluginManager),
    },
  )
}
