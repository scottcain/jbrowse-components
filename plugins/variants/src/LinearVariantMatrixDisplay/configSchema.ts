import { ConfigurationSchema } from '@jbrowse/core/configuration'
import PluginManager from '@jbrowse/core/PluginManager'
import { linearBasicDisplayConfigSchemaFactory } from '@jbrowse/plugin-linear-genome-view'

// locals
import configSchema from '../LinearVariantMatrixRenderer/configSchema'

/**
 * #config LinearVariantMatrixDisplay
 */
function x() {} // eslint-disable-line @typescript-eslint/no-unused-vars

export default function configSchemaF(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'LinearVariantMatrixDisplay',
    {
      renderer: configSchema,
    },
    {
      /**
       * #baseConfiguration
       */
      baseConfiguration: linearBasicDisplayConfigSchemaFactory(pluginManager),
      explicitlyTyped: true,
    },
  )
}
