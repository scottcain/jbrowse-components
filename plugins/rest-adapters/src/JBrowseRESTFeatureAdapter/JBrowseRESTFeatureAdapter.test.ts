import { toArray } from 'rxjs/operators'
import Adapter from './JBrowseRESTFeatureAdapter'
import configSchema from './configSchema'
import PluginManager from '@jbrowse/core/PluginManager'
import { firstValueFrom } from 'rxjs'

function mockFetchResponses(
  adapter: Adapter,
  mockResponses: Record<string, unknown>,
) {
  return (
    jest
      // @ts-ignore
      .spyOn(adapter, 'fetch')
      // @ts-ignore
      .mockImplementation(async (fetcher, url: string, signal) => {
        return {
          async json() {
            if (url in mockResponses) {
              return mockResponses[url]
            }
            throw new Error('no mock for ' + url)
          },
        }
      })
  )
}

function createTestAdapter(
  configuration: Record<string, unknown>,
  mockResponses: Record<string, unknown>,
) {
  const stubManager = new PluginManager()
  const adapter = new Adapter(
    configSchema.create(configuration),
    undefined,
    stubManager,
  )

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const fetchMock: any = mockFetchResponses(adapter, {
    'https://mock.url/mock/url/features/volvox/21?start=34960388&end=35960388&extra_query=42':
      {
        features: testFeatures1,
      },
    'https://mock.url/mock/url/volvox/reference_sequences?extra_query=42': [
      '21',
    ],
  })

  return { adapter, fetchMock }
}

const testFeatures1 = [{ start: 100, end: 200, refName: '21', uniqueId: 'foo' }]

test('adapter can fetch features and stats from mocked API with no region stats', async () => {
  const { adapter, fetchMock } = createTestAdapter(
    {
      location: { uri: 'https://mock.url/mock/url' },
      extra_query: { extra_query: 42 },
      optional_resources: {},
    },
    {
      'https://mock.url/mock/url/features/volvox/21?start=34960388&end=35960388&extra_query=42':
        {
          features: testFeatures1,
        },
      'https://mock.url/mock/url/volvox/reference_sequences?extra_query=42': [
        '21',
      ],
    },
  )

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: '21',
    start: 34960388,
    end: 35960388,
  })

  const featuresArray = await firstValueFrom(features.pipe(toArray()))
  expect(featuresArray.length).toBe(testFeatures1.length)
  expect(featuresArray[0].get('refName')).toBe('21')
  expect(featuresArray[0].id()).toBe('foo')
  // @ts-ignore
  expect(fetchMock.mock.calls[0][1]).toBe(
    'https://mock.url/mock/url/features/volvox/21?start=34960388&end=35960388&extra_query=42',
  )

  expect(
    await adapter.hasDataForRefName('ctgA', { assemblyName: 'volvox' }),
  ).toBe(false)
  expect(
    await adapter.hasDataForRefName('21', { assemblyName: 'volvox' }),
  ).toBe(true)

  expect(
    await adapter.getRegionStats({
      assemblyName: 'volvox',
      refName: '21',
      start: 34960388,
      end: 35960388,
    }),
  ).toMatchSnapshot()
})
