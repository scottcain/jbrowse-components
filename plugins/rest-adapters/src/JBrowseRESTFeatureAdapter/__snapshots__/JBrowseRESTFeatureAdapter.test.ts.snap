// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features and stats from mocked API with no region stats 1`] = `
{
  "basesCovered": 1000001,
  "featureCount": 1,
  "featureDensity": 9.99999000001e-7,
}
`;
